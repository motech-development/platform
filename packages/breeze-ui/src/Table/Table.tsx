import React, { FC, memo, ReactNode } from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  display: block;
  overflow-x: auto;
  width: 100%;
  -webkit-overflow-scrolling: touch;
  -ms-overflow-style: -ms-autohiding-scrollbar;
`;

const BaseTable = styled.table`
  max-width: 100%;
  width: 100%;
`;

export interface ITableProps {
  children: ReactNode;
}

const Table: FC<ITableProps> = ({ children }) => (
  <Wrapper>
    <BaseTable>{children}</BaseTable>
  </Wrapper>
);

export default memo(Table);
