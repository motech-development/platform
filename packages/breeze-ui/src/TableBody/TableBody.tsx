import { memo } from 'react';
import styled from 'styled-components';

const TableBody = styled.tbody`
  background-color: #fff;
  color: #000;
`;

export default memo(TableBody);
