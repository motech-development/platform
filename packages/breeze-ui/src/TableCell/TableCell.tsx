/* eslint-disable react/jsx-props-no-spreading */
import React, { FC, memo, TdHTMLAttributes } from 'react';
import styled from 'styled-components';

const baseStyles = (noWrap: boolean, align = 'inherit') => `
  padding: 10px;
  text-align: ${align};
  vertical-align: middle;
  ${noWrap ? 'white-space: nowrap;' : ''}
`;

interface ITableCell {
  $noWrap: boolean;
}

const TableDataCell = styled.td<ITableCell>`
  ${({ align, $noWrap }) => `
    ${baseStyles($noWrap, align)}
    font-weight: 300;
  `}
`;

const TableHeadCell = styled.th<ITableCell>`
  ${({ align, $noWrap }) => `
    ${baseStyles($noWrap, align)}
    font-family: 'Cabin', sans-serif;
    font-weight: 600;
  `}
`;

export interface ITableCellProps
  extends TdHTMLAttributes<HTMLTableDataCellElement> {
  as?: 'td' | 'th';
  noWrap?: boolean;
}

const TableCell: FC<ITableCellProps> = ({
  as = 'td',
  noWrap = true,
  ...rest
}) =>
  as === 'td' ? (
    <TableDataCell $noWrap={noWrap} {...rest} />
  ) : (
    <TableHeadCell $noWrap={noWrap} {...rest} />
  );

export default memo(TableCell);
