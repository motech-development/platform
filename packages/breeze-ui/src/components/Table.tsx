import { FC, HTMLProps, ReactNode } from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  border-bottom: 2px solid #007fa8;
  display: block;
  overflow-x: auto;
  width: 100%;
  -webkit-overflow-scrolling: touch;
  -ms-overflow-style: -ms-autohiding-scrollbar;
`;

interface IBaseTable {
  $fixed: boolean;
}

const BaseTable = styled.table<IBaseTable>`
  ${({ $fixed }) => `
    background-color: #fff;
    max-width: 100%;
    table-layout: ${$fixed ? 'fixed' : 'auto'};
    width: 100%;
  `}
`;

export interface ITableProps extends HTMLProps<HTMLTableElement> {
  children: ReactNode;
  fixed?: boolean;
}

const Table: FC<ITableProps> = ({ children, fixed = false, ...rest }) => (
  <Wrapper>
    {/* eslint-disable-next-line react/jsx-props-no-spreading */}
    <BaseTable $fixed={fixed} {...rest}>
      {children}
    </BaseTable>
  </Wrapper>
);

export default Table;
