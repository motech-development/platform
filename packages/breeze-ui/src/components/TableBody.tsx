import { FC, HTMLProps } from 'react';
import { classNames } from '../utils/className';

export type TTableBodyProps = HTMLProps<HTMLTableSectionElement>;

const TableBody: FC<TTableBodyProps> = ({ className = '', ...rest }) => (
  <tbody className={classNames('font-sans text-sm', className)} {...rest} />
);

export default TableBody;
