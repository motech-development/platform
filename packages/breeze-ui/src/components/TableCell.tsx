import { FC, forwardRef, ThHTMLAttributes } from 'react';
import { Box } from 'react-polymorphic-box';
import TAlignment from '../utils/alignment';
import { classNames, textAlignmentClass } from '../utils/className';

type TTableCellAs = 'td' | 'th';

export interface ITableCellProps
  extends ThHTMLAttributes<HTMLTableDataCellElement> {
  align?: TAlignment;
  as?: TTableCellAs;
  noWrap?: boolean;
}

const TableCell: FC<ITableCellProps> = forwardRef<
  HTMLTableDataCellElement,
  ITableCellProps
>(
  (
    { align = 'left', as = 'td', className = '', noWrap = true, ...rest },
    ref,
  ) => (
    <Box
      as={as}
      className={classNames(
        'px-6 py-3',
        className,
        as === 'td' ? 'bg-white bg-opacity-70' : '',
        noWrap ? 'whitespace-nowrap' : '',
        textAlignmentClass(align),
      )}
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    />
  ),
);

export default TableCell;
