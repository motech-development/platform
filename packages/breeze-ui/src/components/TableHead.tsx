import { FC, HTMLProps } from 'react';
import { classNames } from '../utils/className';

export type TTableHeadProps = HTMLProps<HTMLTableSectionElement>;

const TableHead: FC<TTableHeadProps> = ({ className = '', ...rest }) => (
  <thead
    className={classNames('font-display font-semibold', className)}
    {...rest}
  />
);

export default TableHead;
