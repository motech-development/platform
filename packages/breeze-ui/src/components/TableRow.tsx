import { FC, forwardRef, TableHTMLAttributes } from 'react';
import { classNames, themeClass } from '../utils/className';
import TTheme from '../utils/theme';

export interface ITableRowProps
  extends TableHTMLAttributes<HTMLTableRowElement> {
  colour?: TTheme;
}

const TableRow: FC<ITableRowProps> = forwardRef<
  HTMLTableRowElement,
  ITableRowProps
>(({ className = '', colour = 'secondary', ...rest }, ref) => (
  <>
    {/* @tailwind: bg-blue-700 bg-gray-100 bg-green-700 bg-red-700 bg-yellow-700 */}
    {/* @tailwind: border-blue-800 border-gray-200 border-green-800 border-red-800 border-yellow-800 */}
    {/* @tailwind: text-blue-100 text-gray-700 text-green-100 text-red-100 text-yellow-100 */}
    <tr
      className={classNames(
        'border-b-2',
        className,
        themeClass(
          colour,
          'bg-{theme}-700 border-{theme}-800 text-{theme}-100',
          {
            secondary: 'bg-{theme}-100 border-{theme}-200 text-{theme}-700',
          },
        ),
      )}
      ref={ref}
      {...rest}
    />
  </>
));

export default TableRow;
