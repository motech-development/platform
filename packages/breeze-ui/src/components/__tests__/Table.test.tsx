import { render, RenderResult } from '@testing-library/react';
import TableBody from '../TableBody';
import TableCell from '../TableCell';
import TableHead from '../TableHead';
import TableRow from '../TableRow';
import Table from '../Table';

describe('Table', () => {
  let component: RenderResult;

  describe('when not fixed', () => {
    beforeEach(() => {
      component = render(
        <Table>
          <TableHead data-testid="table-head">
            <TableRow colour="primary">
              <TableCell as="th" colSpan={2}>
                My table
              </TableCell>
            </TableRow>
          </TableHead>

          <TableBody data-testid="table-body">
            <TableRow colour="primary" data-testid="primary-row">
              <TableCell as="th" colSpan={2}>
                16th April 2019
              </TableCell>
            </TableRow>

            <TableRow colour="secondary" data-testid="secondary-row">
              <TableCell as="th" data-testid="table-cell-header">
                Transaction
              </TableCell>
              <TableCell as="th">Value</TableCell>
            </TableRow>

            <TableRow data-testid="default-row">
              <TableCell>Payment from Joe Bloggs</TableCell>
              <TableCell data-testid="table-cell" align="left">
                £150
              </TableCell>
            </TableRow>

            <TableRow>
              <TableCell>Lunch</TableCell>
              <TableCell>-£5.25</TableCell>
            </TableRow>
          </TableBody>
        </Table>,
      );
    });

    it('should render the table correctly', () => {
      const { container } = component;

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('when fixed', () => {
    beforeEach(() => {
      component = render(
        <Table fixed>
          <TableHead data-testid="table-head">
            <TableRow colour="primary">
              <TableCell as="th" colSpan={2}>
                My table
              </TableCell>
            </TableRow>
          </TableHead>

          <TableBody data-testid="table-body">
            <TableRow colour="primary" data-testid="primary-row">
              <TableCell as="th" colSpan={2}>
                16th April 2019
              </TableCell>
            </TableRow>

            <TableRow colour="secondary" data-testid="secondary-row">
              <TableCell as="th" data-testid="table-cell-header">
                Transaction
              </TableCell>
              <TableCell as="th">Value</TableCell>
            </TableRow>

            <TableRow data-testid="default-row">
              <TableCell>Payment from Joe Bloggs</TableCell>
              <TableCell data-testid="table-cell" align="left">
                £150
              </TableCell>
            </TableRow>

            <TableRow>
              <TableCell>Lunch</TableCell>
              <TableCell>-£5.25</TableCell>
            </TableRow>
          </TableBody>
        </Table>,
      );
    });

    it('should render the table correctly', () => {
      const { container } = component;

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
