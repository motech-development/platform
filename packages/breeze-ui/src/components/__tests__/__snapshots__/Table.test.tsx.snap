// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table when fixed should render the table correctly 1`] = `
<div
  class="flex flex-col"
>
  <div
    class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
  >
    <div
      class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
    >
      <div
        class="shadow overflow-hidden border-b border-gray-200"
      >
        <table
          class="min-w-full table-fixed"
        >
          <thead
            class="font-display font-semibold"
            data-testid="table-head"
          >
            <tr
              class="border-b-2 bg-blue-700 border-blue-800 text-blue-100"
            >
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
                colspan="2"
              >
                My table
              </th>
            </tr>
          </thead>
          <tbody
            class="font-sans text-sm"
            data-testid="table-body"
          >
            <tr
              class="border-b-2 bg-blue-700 border-blue-800 text-blue-100"
              data-testid="primary-row"
            >
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
                colspan="2"
              >
                16th April 2019
              </th>
            </tr>
            <tr
              class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
              data-testid="secondary-row"
            >
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
                data-testid="table-cell-header"
              >
                Transaction
              </th>
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
              >
                Value
              </th>
            </tr>
            <tr
              class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
              data-testid="default-row"
            >
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
              >
                Payment from Joe Bloggs
              </td>
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
                data-testid="table-cell"
              >
                £150
              </td>
            </tr>
            <tr
              class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
            >
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
              >
                Lunch
              </td>
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
              >
                -£5.25
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table when not fixed should render the table correctly 1`] = `
<div
  class="flex flex-col"
>
  <div
    class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
  >
    <div
      class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
    >
      <div
        class="shadow overflow-hidden border-b border-gray-200"
      >
        <table
          class="min-w-full table-auto"
        >
          <thead
            class="font-display font-semibold"
            data-testid="table-head"
          >
            <tr
              class="border-b-2 bg-blue-700 border-blue-800 text-blue-100"
            >
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
                colspan="2"
              >
                My table
              </th>
            </tr>
          </thead>
          <tbody
            class="font-sans text-sm"
            data-testid="table-body"
          >
            <tr
              class="border-b-2 bg-blue-700 border-blue-800 text-blue-100"
              data-testid="primary-row"
            >
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
                colspan="2"
              >
                16th April 2019
              </th>
            </tr>
            <tr
              class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
              data-testid="secondary-row"
            >
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
                data-testid="table-cell-header"
              >
                Transaction
              </th>
              <th
                class="px-6 py-3 whitespace-nowrap text-left"
              >
                Value
              </th>
            </tr>
            <tr
              class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
              data-testid="default-row"
            >
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
              >
                Payment from Joe Bloggs
              </td>
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
                data-testid="table-cell"
              >
                £150
              </td>
            </tr>
            <tr
              class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
            >
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
              >
                Lunch
              </td>
              <td
                class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
              >
                -£5.25
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
